### ISSUE TITLE
Preserving Tab key behavior in the monaco editor

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#48126]
### Describe the Issue

The default behavior for the Tab key in the monaco editor is to add a tab (or remove a tab with `Shift + Tab`). But this effectively creates a Tab trap in the editor. The solution is to use `Ctrl + M` to change the behavior of the Tab key so that it instead moves focus to the next focusable element on the page (or previous focusable element for `Shift + Tab`). As an alternative, you can also use `Esc` to move focus out of the editor, but this moves the focus to a wrapper div around `main` and so you will only be able to Tab to focusable elements before the editor but you will never be able to Tab to focusable elements after the editor without using `Ctrl + M`.

However, the `Ctrl + M` solution does not always persist. For example, if you reload the page then the default Tab behavior is reinstated. Because keyboard only users (at least those who aren't using screen readers) will depend entirely on the Tab key to access all focusable elements and will need to use the `Ctrl + M` solution to reach focusable elements after the editor, I think it would be nice if we provided a way to make the user's chosen Tab behavior permanent. On a side note, monaco already provides keyboard shortcuts to format the code in the editor, so it is not necessary for keyboard only users to switch back to the default Tab behavior in order to add leading tabs.

### Affected Page

Any page with the monaco editor

### Steps to Reproduce

1. Using the keyboard, hit the Tab key until the focus is in the monaco editor (if it isn't already).
2. Hit the Tab key while focus is in the editor and notice that it adds tabs.
3. Hit `Ctrl + M` once to change the Tab behavior so that it will now move focus to the next focusable element.
4. Hit `Tab` and notice that the focus moves to the next focusable element.
5. Reload the page and repeat steps 1 and 2. Notice that the default Tab behavior is back.

### Expected behavior

I have already implemented a solution to save the user's Tab preference (and have a PR ready to go) by creating a new keyboard shortcut that not only toggles between the two Tab options (just like `Ctrl + M` does) but also saves the preference to localStorage so that it can persist. My original thought was that it would be better to provide a different shortcut and allow `Ctrl + M` to work as it always had. But now I'm not so sure and perhaps we should just hijack `Ctrl + M` so it saves the user's preference. My thinking was that if someone accidentally hit `Ctrl + M` and didn't know how to change it back they could at least reload the page to get the default Tab behavior back. But now I'm thinking this will probably not happen very often and it's not worth the possible confusion of having another keyboard shortcut that does the same thing as `Ctrl + M` with the addition that it makes the preference permanent.

I'd be interested to know which approach you all think is better:

A) Modify the existing `Ctrl + M` Tab toggle shortcut so that it saves the user's preference so we can make it persist.
B) Add a new Tab toggle shortcut that mimics `Ctrl + M` and also saves the user's preference so we can make it persist. In this solution, the original `Ctrl + M` would still work as normal.

Or perhaps someone has a C solution to offer?

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

**Comment by naomi-lgbt** [comment]
> A) Modify the existing Ctrl + M Tab toggle shortcut so that it saves the user's preference so we can make it persist.

This is what makes the most sense to me, but I'm also not a keyboard-only user. But generally, when I toggle something I expect it to stay toggled until I toggle it again.

**Comment by ojeytonwilliams** [comment]
> My thinking was that if someone accidentally hit Ctrl + M and didn't know how to change it back they could at least reload the page to get the default Tab behavior back. 

i.e. if they accidentally hit Ctrl + M without noticing?  I think that will be rare enough that we don't have to worry.  As long as Ctrl + M toggles back and forth between the two modes, that should be solid.

**Comment by bbsmooth** [comment]
As someone who is constantly using this site with the keyboard only (for accessibility testing) I find it very annoying to have to keep doing `Ctrl + m`. Basically, I never want the Tab key to add tabs in the editor, and making this permanent has been on my list of things to do for quite some time now. Of course, I may not be a "typical" user of this site, but I do think that others who use this site with the keyboard only probably don't want the default Tab key behavior either and will appreciate this small change.

As far as implementation goes, hijacking the current `Ctrl + m` hotkey is proving to be much tougher than anticipated. Monaco does not provide a way to alter/extend the default behavior. And you can't remove it and replace it with your own implementation because it is built into the accessibility pop-up menu (the one you get when you do `Alt + F1`). So I'm guessing I'll have to add a separate keyup listener for `Ctrl + m` and do the localStorage management there.

**Comment by ojeytonwilliams** [comment]
@bbsmooth have you tried playing with `editor._standaloneKeybindingService.addDynamicKeybinding`? It might do what you need it to. 

Eventually https://microsoft.github.io/monaco-editor/api/modules/monaco.editor.html#addKeybindingRule could be the way to go, but we can't currently update the editor to a version that includes this function.

**Comment by LaloQuera710** [comment]
[[[]()]()]()

**Comment by LaloQuera710** [comment]
[***]()
